File: python/src/threads/hooks/extension.py
"""`extension()` (spec/api.json): the one extension primitive. Trusted host code,
not a security boundary. Hooks and observers come only through it."""

import re
from collections.abc import Awaitable, Callable, Mapping, Sequence
from dataclasses import dataclass, field
from typing import Final, Required, TypedDict, Unpack

from pydantic import JsonValue
from typing_extensions import TypeVar

from threads.agents.bindings import AppTool
from threads.agents.config import ConfigError
from threads.agents.context import RunContext
from threads.hooks.runner import Bound, Call, HookRunner
from threads.hooks.types import HookName, Hooks, wire_name
from threads.log import Event, JsonObject, Origin, ToolSpec
from threads.loop.model import LookupResult
from threads.loop.tools import Dispatched

type Observer = Callable[[Event], Awaitable[None]]

D = TypeVar("D", default=None)
"""The deps an extension's hooks and tools read. PEP 696's default (typing_extensions, which
pydantic already depends on): an extension with no hooks or tools is Extension[None], never
Extension[Unknown]."""

_NAME: Final = re.compile(r"[a-z][a-z0-9_]{0,63}")
DEFAULT_TIMEOUT_MS: Final = 5000


class ExtensionOptions[D](TypedDict, total=False):
    name: Required[str]
    tools: Sequence[AppTool[D]]
    """Namespaced `<name>__<tool>`. Like hooks, they get the run's deps."""
    instructions: str
    hooks: Hooks[D]
    on: Mapping[str, Observer]
    setup: Callable[[], Awaitable[None]]
    hook_timeout_ms: int


# No slots: the setup memory holds an extension weakly, and before 3.12.4 a slotted generic
# dataclass drops its weakref slot (CPython gh-118033), so every run failed setup.
@dataclass(frozen=True)
class Extension[D]:
    """spec/api.json `Extension`. Build it with `extension()`. D is the deps its hooks and tools
    read: the agent's Deps, None when they read none."""

    name: str
    instructions: str = ""
    hooks: Hooks[D] = field(default_factory=Hooks[D])
    on: Mapping[str, Observer] = field(default_factory=dict[str, Observer])
    setup: Callable[[], Awaitable[None]] | None = None
    hook_timeout_ms: int = DEFAULT_TIMEOUT_MS
    tools: tuple[AppTool[D], ...] = ()

    def manifest(self) -> JsonValue:
        """What the config pin covers: which hooks this extension declares, in wire-enum order,
        its observers, sorted, and its timeout. Code is identified by the host, not the
        log; the pin makes a changed hook set a changed config."""
        hooks: list[JsonValue] = [wire_name(h) for h in _names(self.hooks)]
        observers: list[JsonValue] = [*sorted(self.on)]
        return {
            "name": self.name,
            "hooks": hooks,
            "observers": observers,
            "hook_timeout_ms": self.hook_timeout_ms,
        }


@dataclass(frozen=True, slots=True)
class Namespaced[D]:
    """An extension's tool under its pinned `<ext>__<tool>` name."""

    name: str
    tool: AppTool[D]
    extension: str

    def spec(self) -> ToolSpec:
        # Not model-visible: line 0 leaves the origin out, config_hash covers it (drift reads it).
        origin = Origin(extension=self.extension)
        return self.tool.spec().model_copy(update={"name": self.name, "origin": origin})

    def inner(self) -> object:
        """The extension's own tool, whatever deps it reads (what deferral looks at)."""
        return self.tool

    def invalid(self, input: JsonObject) -> str | None:
        return self.tool.invalid(input)

    async def run(self, input: JsonObject, ctx: RunContext[D]) -> Dispatched:
        return await self.tool.run(input, ctx)

    async def lookup(self, effect_key: str, ctx: RunContext[D]) -> LookupResult[str]:
        return await self.tool.lookup(effect_key, ctx)


def extension_tools[D](extensions: Sequence[Extension[D]]) -> tuple[Namespaced[D], ...]:
    """Every extension's tools, namespaced and sorted by that name."""
    tools = [Namespaced(f"{e.name}__{t.name}", t, e.name) for e in extensions for t in e.tools]
    return tuple(sorted(tools, key=lambda t: t.name))


def _names[D](hooks: Hooks[D]) -> list[HookName]:
    return [name for name in _ALL if name in hooks]


_ALL: Final[tuple[HookName, ...]] = (
    "session_start",
    "session_end",
    "before_input",
    "before_model",
    "after_model",
    "before_tool",
    "permission_request",
    "permission_denied",
    "after_tool",
    "before_tool_result",
    "after_tool_batch",
    "before_compact",
    "after_compact",
    "on_stop",
    "on_stop_failure",
    "subagent_start",
    "subagent_stop",
    "before_model_switch",
    "after_model_switch",
    "notification",
)


def extension(**options: Unpack[ExtensionOptions[D]]) -> Extension[D]:
    """spec/api.json `extension`. Pure. Raises ConfigError for a name the wire can't hold or a
    non-positive timeout."""
    name = options["name"]
    if _NAME.fullmatch(name) is None:
        raise ConfigError("invalid_config", f"extension name {name!r} is not a wire Name")
    timeout = options.get("hook_timeout_ms", DEFAULT_TIMEOUT_MS)
    if timeout <= 0:
        raise ConfigError("invalid_config", f"{name}: hook_timeout_ms must be positive")
    return Extension(
        name,
        options.get("instructions", ""),
        options.get("hooks", Hooks[D]()),
        dict(options.get("on", {})),
        options.get("setup"),
        timeout,
        tuple(options.get("tools", ())),
    )


def bind[D](extensions: Sequence[Extension[D]], ctx: RunContext[D]) -> HookRunner:
    """The run's hooks, each with its context bound (spec/api.json `RunContext`)."""
    return HookRunner(
        [
            Bound(
                e.name,
                e.hook_timeout_ms,
                {name: _with(fn, ctx) for name in _ALL if (fn := e.hooks.get(name)) is not None},
            )
            for e in extensions
        ]
    )


def _with(fn: Call, ctx: object) -> Call:
    async def call(*args: object) -> object:
        return await fn(*args, ctx)

    return call

File: typescript/packages/core/src/agent/extension.ts
import type { EventOf } from "../fold/state";
import type { Observer, ObserverHandler } from "../hooks/observers";
import type {
  HookContext,
  HookName,
  HookReturns,
  LoopExtension,
  LoopHooks,
  SessionSource,
} from "../hooks/types";
import { type KnownEvent, Name } from "../log";
import type { RunErrorCode } from "../loop";
import type { ReducedState } from "../reduce";
import { ConfigError } from "./errors";
import type { RunContext, Tool } from "./tool";

// extension() (spec/api.json): the one primitive for tools, trusted
// instructions, hooks and observers. Trusted host code, not a security boundary. What it pins
// (name, instructions, tools, which hooks and observers exist, the timeout) is part of the
// thread's config_hash, so no agent path and no later run can change a thread's hooks.

type Call = EventOf<"tool_call">["data"];
type Result = EventOf<"tool_result">["data"];
type Response = EventOf<"model_response" | "model_response_recovered">["data"];
type Ctx<Deps> = RunContext<Deps>;
type Settings = EventOf<"settings_changed">["data"]["settings"];

/** spec/api.json types.Hooks, in TS casing. Every return is a wire hook_decision value. */
export type Hooks<Deps = undefined> = {
  readonly sessionStart?: (
    source: SessionSource,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["session_start"]>;
  readonly sessionEnd?: (ctx: Ctx<Deps>) => Promise<void>;
  readonly beforeInput?: (
    input: EventOf<"user_input">["data"],
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["before_input"]>;
  readonly beforeModel?: (
    state: ReducedState,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["before_model"]>;
  readonly afterModel?: (
    state: ReducedState,
    response: Response,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["after_model"]>;
  readonly beforeTool?: (
    call: Call,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["before_tool"]>;
  readonly permissionRequest?: (
    call: Call,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["permission_request"]>;
  readonly permissionDenied?: (call: Call, ctx: Ctx<Deps>) => Promise<void>;
  readonly afterTool?: (
    call: Call,
    result: Result,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["after_tool"]>;
  readonly beforeToolResult?: (
    call: Call,
    result: Result,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["before_tool_result"]>;
  readonly afterToolBatch?: (
    state: ReducedState,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["after_tool_batch"]>;
  readonly beforeCompact?: (
    state: ReducedState,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["before_compact"]>;
  readonly afterCompact?: (
    state: ReducedState,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["after_compact"]>;
  readonly onStop?: (
    state: ReducedState,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["on_stop"]>;
  readonly onStopFailure?: (
    code: RunErrorCode,
    ctx: Ctx<Deps>,
  ) => Promise<void>;
  readonly subagentStart?: (
    call: Call,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["subagent_start"]>;
  readonly subagentStop?: (
    finished: EventOf<"agent_finished">["data"],
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["subagent_stop"]>;
  readonly beforeModelSwitch?: (
    settings: Settings,
    ctx: Ctx<Deps>,
  ) => Promise<HookReturns["before_model_switch"]>;
  readonly afterModelSwitch?: (
    settings: Settings,
    ctx: Ctx<Deps>,
  ) => Promise<void>;
  readonly notification?: (event: KnownEvent, ctx: Ctx<Deps>) => Promise<void>;
};

export type ExtensionOptions<Deps = undefined> = {
  /** Unique; its tools are namespaced `<name>__<tool>`. */
  readonly name: string;
  readonly tools?: readonly Tool<unknown, unknown, Deps>[];
  /** Static; line 0 after the base instructions, in declaration order. */
  readonly instructions?: string;
  readonly hooks?: Hooks<Deps>;
  /** Observers keyed by event type, or "*": committed events after append. */
  readonly on?: Readonly<Record<string, ObserverHandler>>;
  /** Runs once at check() or the first run. Throwing is a ConfigError. */
  readonly setup?: () => Promise<void>;
  /** Per hook call; default 5 s. */
  readonly hookTimeoutMs?: number;
};

export type Extension<Deps = undefined> = ExtensionOptions<Deps> & {
  readonly kind: "extension";
};

const DEFAULT_TIMEOUT_MS = 5000;

export function extension<Deps = undefined>(
  options: ExtensionOptions<Deps>,
): Extension<Deps> {
  if (!Name.safeParse(options.name).success)
    throw new ConfigError(
      "invalid_config",
      `extension name ${JSON.stringify(options.name)} is not a valid name`,
    );
  const timeout = options.hookTimeoutMs;
  if (timeout !== undefined && (!Number.isInteger(timeout) || timeout <= 0))
    throw new ConfigError(
      "invalid_config",
      `extension ${options.name}: hookTimeoutMs must be a positive integer`,
    );
  return { ...options, kind: "extension" };
}

/** The wire name of every hook an extension defines, in the order of the wire enum. */
export function hookNames<Deps>(ext: Extension<Deps>): readonly HookName[] {
  return Object.keys(bindHooks(ext, () => undefined)).filter(isHookName);
}

const HOOK_NAMES: ReadonlySet<string> = new Set<HookName>([
  "session_start",
  "session_end",
  "before_input",
  "before_model",
  "after_model",
  "before_tool",
  "permission_request",
  "permission_denied",
  "after_tool",
  "before_tool_result",
  "after_tool_batch",
  "before_compact",
  "after_compact",
  "on_stop",
  "stop_failure",
  "subagent_start",
  "subagent_stop",
  "before_model_switch",
  "after_model_switch",
  "notification",
]);

const isHookName = (name: string): name is HookName => HOOK_NAMES.has(name);

/** The loop's view of an extension, bound to this run's context. */
export function loopExtension<Deps>(
  ext: Extension<Deps>,
  ctx: (hook: HookContext) => Ctx<Deps>,
): LoopExtension {
  return {
    name: ext.name,
    timeoutMs: ext.hookTimeoutMs ?? DEFAULT_TIMEOUT_MS,
    hooks: bindHooks(ext, ctx),
  };
}

export function observerOf<Deps>(ext: Extension<Deps>): Observer | undefined {
  return ext.on === undefined ? undefined : { name: ext.name, on: ext.on };
}

type Bind<Deps> = (hook: HookContext) => Ctx<Deps> | undefined;

/** Each camelCase hook as the loop calls it: wire name, args tuple, the run context bound. */
function bindHooks<Deps>(ext: Extension<Deps>, bind: Bind<Deps>): LoopHooks {
  const h = ext.hooks ?? {};
  const c = (hc: HookContext): Ctx<Deps> => {
    const bound = bind(hc);
    if (bound === undefined) throw new Error("hook names only, not callable");
    return bound;
  };
  const one = <K extends HookName>(
    name: K,
    fn: NonNullable<LoopHooks[K]> | undefined,
  ): LoopHooks => (fn === undefined ? {} : { [name]: fn });
  const {
    sessionStart,
    sessionEnd,
    beforeInput,
    beforeModel,
    afterModel,
    beforeTool,
    permissionRequest,
    permissionDenied,
    afterTool,
    beforeToolResult,
    afterToolBatch,
    beforeCompact,
    afterCompact,
    onStop,
    onStopFailure,
    subagentStart,
    subagentStop,
    beforeModelSwitch,
    afterModelSwitch,
    notification,
  } = h;
  const opt = <F, K extends HookName>(
    fn: F | undefined,
    name: K,
    wrap: (f: F) => NonNullable<LoopHooks[K]>,
  ): LoopHooks => one(name, fn === undefined ? undefined : wrap(fn));
  return {
    ...opt(
      sessionStart,
      "session_start",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(sessionEnd, "session_end", (f) => (_, x) => f(c(x))),
    ...opt(
      beforeInput,
      "before_input",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      beforeModel,
      "before_model",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      afterModel,
      "after_model",
      (f) =>
        ([a, b], x) =>
          f(a, b, c(x)),
    ),
    ...opt(
      beforeTool,
      "before_tool",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      permissionRequest,
      "permission_request",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      permissionDenied,
      "permission_denied",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      afterTool,
      "after_tool",
      (f) =>
        ([a, b], x) =>
          f(a, b, c(x)),
    ),
    ...opt(
      beforeToolResult,
      "before_tool_result",
      (f) =>
        ([a, b], x) =>
          f(a, b, c(x)),
    ),
    ...opt(
      afterToolBatch,
      "after_tool_batch",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      beforeCompact,
      "before_compact",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      afterCompact,
      "after_compact",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      onStop,
      "on_stop",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      onStopFailure,
      "stop_failure",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      subagentStart,
      "subagent_start",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      subagentStop,
      "subagent_stop",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      beforeModelSwitch,
      "before_model_switch",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      afterModelSwitch,
      "after_model_switch",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
    ...opt(
      notification,
      "notification",
      (f) =>
        ([a], x) =>
          f(a, c(x)),
    ),
  };
}
